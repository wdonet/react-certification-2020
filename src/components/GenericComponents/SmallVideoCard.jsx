import React, { useContext, useState } from 'react';
import styled from 'styled-components';
import AppContext from '../../providers/AppContext';
import { Button } from '../../ui'

const StyledSmallVideoCard = styled.div`
  display: flex;
  padding: 4px;
  margin: 4px;
  border-radius: 4px;
  border-bottom: 1px solid ${({theme}) => theme.color.secondary };
  color: ${({theme}) => theme.color.fontSecondary };
  background: ${({theme}) => theme.color.surface };
  cursor: pointer;
`;

const StyledSmallVideoCardDescription = styled.div`
  height: 100%;
  -webkit-box-flex: 1;
  flex-grow: 1;
  box-sizing: border-box;
`;

const StyledCaption = styled.div`
  display: block;
  height: fit-content;
  padding: 4px;
`;

const FAVORITES_KEY = "favorites";
const getParsedFavorites = () => JSON.parse(window.localStorage.getItem(FAVORITES_KEY));

const SmallVideoCard = ({ video, onClick, onFavorite }) => {

  const getLabel = (videoId) => 
    getParsedFavorites() && getParsedFavorites()[videoId] 
    ? "Remove favorite"
    : "Add favorite";

  const { theme } = useContext(AppContext);
  const { thumbnails, title, description } = video.snippet;
  const [buttonLabel, setButtonLabel] = useState(getLabel(video.id.videoId));

  const updateLabel = () => { setButtonLabel(getLabel(video.id.videoId)); }

  return (
    <StyledSmallVideoCard 
      theme={theme} 
      onClick={ ({target}) => !target.matches("button") && onClick && onClick() }
    >
      <img src={thumbnails.default.url} alt={title} />
      <StyledCaption>
        <StyledSmallVideoCardDescription>{description}</StyledSmallVideoCardDescription>
        <Button 
          data-testid="caption-add-favorite" 
          onClick={() => { onFavorite && onFavorite(video, updateLabel) }}
        >
          {buttonLabel}
        </Button>
      </StyledCaption>
    </StyledSmallVideoCard>
  );
};

export default SmallVideoCard;
