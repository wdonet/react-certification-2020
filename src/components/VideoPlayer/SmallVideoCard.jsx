import React from 'react';
import styled from 'styled-components';

const StyledSmallVideoCard = styled.div`
  display: flex;
  padding: 4px;
  margin: 4px;
  border-bottom: 1px solid #eaeaea;
  cursor: pointer;
`;

const StyledSmallVideoCardDescription = styled.div`
  height: 100%;
  -webkit-box-flex: 1;
  flex-grow: 1;
  padding: 5px;
  box-sizing: border-box;
`;

const SmallVideoCard = ({ video, onClick }) => {
  const { thumbnails, title, description } = video.snippet;
  return (
    <StyledSmallVideoCard onClick={() => onClick && onClick()}>
      <img src={thumbnails.default.url} alt={title} />
      <StyledSmallVideoCardDescription>{description}</StyledSmallVideoCardDescription>
    </StyledSmallVideoCard>
  );
};

export default SmallVideoCard;
