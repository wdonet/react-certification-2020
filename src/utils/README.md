# Utils

Set of tools or utils that can be used anywhere in the application, these utils should be pure functions.

## List of Utils

| Util | Description |
| ---- | ----------- |
|      |             |
