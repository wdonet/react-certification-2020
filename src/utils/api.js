export const getVideos = async () => {
  return window.gapi.client.request({
    path: 'https://www.googleapis.com/youtube/v3/videos',
    params: { chart: 'mostPopular', part: 'snippet,statistics', maxResults: 25 },
  });
};

export const searchVideos = async (text) => {
  return window.gapi.client.request({
    path: 'https://www.googleapis.com/youtube/v3/search',
    params: {
      q: text,
      chart: 'mostPopular',
      part: 'snippet',
      maxResults: 25,
      type: 'video',
    },
  });
};

export const getRelatedVideos = async (videoId) => {
  return window.gapi.client.request({
    path: 'https://www.googleapis.com/youtube/v3/search',
    params: {
      part: 'snippet',
      type: 'video',
      maxResults: 25,
      relatedToVideoId: videoId,
    },
  });
};

export const getVideoStatistics = async (videoId) => {
  return window.gapi.client.request({
    path: 'https://www.googleapis.com/youtube/v3/videos',
    params: {
      part: 'snippet,statistics',
      id: videoId,
    },
  });
};
