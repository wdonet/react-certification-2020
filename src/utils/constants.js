export const AUTH_STORAGE_KEY = 'wa_cert_authenticated';
export const DARK_THEME = 'dark_theme_on';
