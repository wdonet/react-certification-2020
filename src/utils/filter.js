function filterByYear(videoList, year) {
  console.log(`videoList: ${videoList}`)
  if (year === '') {
    console.log('returns same videolist');
    return videoList;
  }
  
  let newVidList =  videoList.filter((video) => {
    return video.snippet.publishTime.indexOf(year) >= 0;
  });

  console.log(`newVidList: ${newVidList}`)
  console.log('returns filtered  videolist');
  return newVidList;

}

export default filterByYear;
