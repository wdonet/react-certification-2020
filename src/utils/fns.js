const random = (limit) => {
  return Math.floor(Math.random() * limit);
};

const toggle = (state) => {
  return !state;
};

const shortenCount = (count) => {
  let shortCount = Number(count);
  let pos = 0;
  const units = ['', 'K', 'M', 'B'];

  while (shortCount >= 999) {
    shortCount /= 1000;
    pos += 1;
  }

  return Number(shortCount.toPrecision(2)) + units[pos];
};

export { random, shortenCount, toggle };
