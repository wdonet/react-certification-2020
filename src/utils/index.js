export { default as contextWrapper } from './testing/helpers/contextWrapper';
export { default as renderWithTheme } from './testing/helpers/renderWithTheme';
export { default as googleMockedAPIObject } from './testing/mocks/googleMockedAPIObject';
export { default as youtubeMockedData } from './testing/mocks/youtubeMockedData';
export { default as YTMockedObject } from './testing/mocks/YTMockedObject';
