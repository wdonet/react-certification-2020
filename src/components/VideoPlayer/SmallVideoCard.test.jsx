import React from 'react';
import 'jest-styled-components';
import { getByRole } from '@testing-library/dom';
import { renderWithTheme, youtubeMockedData } from '../../utils';
import SmallVideoCard from './SmallVideoCard';

const build = (Component = <SmallVideoCard />) => {
  const { container } = renderWithTheme(Component);
  return { container };
};

describe('SmallVideoCard', () => {
  it('shows video info', () => {
    const mockVideo = youtubeMockedData.items[0];
    const { firstChild } = build(<SmallVideoCard video={mockVideo} />).container;
    expect(firstChild).toMatchSnapshot();
  });

  it('applies default theme', () => {
    const mockVideo = youtubeMockedData.items[0];
    const { firstChild } = build(<SmallVideoCard video={mockVideo} />).container;
    expect(firstChild).toHaveStyle('display: flex');
    expect(firstChild).toHaveStyle('padding: 4px');
    expect(firstChild).toHaveStyle('margin: 4px');
    expect(firstChild).toHaveStyle('cursor: pointer');
    expect(firstChild).toHaveStyle('border-bottom: 1px solid #eaeaea');
  });

  it('shows video info', () => {
    const { snippet } = youtubeMockedData.items[0];
    const { container } = build(<SmallVideoCard video={{ snippet }} />);
    const image = getByRole(container, 'img');

    expect(image).toHaveAttribute('alt', snippet.title);
    expect(image).toHaveAttribute('src', snippet.thumbnails.default.url);
    expect(container).toHaveTextContent(snippet.description);
  });
});
