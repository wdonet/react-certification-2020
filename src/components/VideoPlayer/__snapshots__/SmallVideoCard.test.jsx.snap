// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SmallVideoCard shows video info 1`] = `
<div
  class="sc-bdVaJa emnpMO"
>
  <img
    alt="Video Tour | Welcome to Wizeline Guadalajara"
    src="https://i.ytimg.com/vi/nmXMgqjQzls/default.jpg"
  />
  <div
    class="sc-bwzfXH kuXOth"
  >
    Follow Hector Padilla, Wizeline Director of Engineering, for a lively tour of our office. In 2018, Wizeline opened its stunning new office in Guadalajara, Jalisco, ...
  </div>
</div>
`;
